src/app/qgisapp.cpp

```diff
@@ -7675,41 +7339,111 @@ bool QgisApp::openLayer( const QString &fileName, bool allowInteractive )
     }
   }
 
-  // try to load it as raster
-  if ( QgsRasterLayer::isValidRasterFileName( fileName ) )
+  QList< QgsProviderSublayerModel::NonLayerItem > nonLayerItems;
+  if ( QgsProjectStorage *ps = QgsApplication::projectStorageRegistry()->projectStorageFromUri( fileName ) )
   {
-    // open .adf as a directory
-    if ( fileName.endsWith( QLatin1String( ".adf" ), Qt::CaseInsensitive ) )
+    const QStringList projects = ps->listProjects( fileName );
```

Comment:
is `ps` guaranteed to be not null?

Reply:
Yep, it's checked a couple of lines above